synapse/federation/federation_client.py

```diff
@@ -1152,23 +1153,30 @@ async def _execute(pdu: EventBase) -> None:
                     % (auth_chain_create_events,)
                 )
 
-            if response.members_omitted and not response.servers_in_room:
-                raise InvalidResponseError(
-                    "members_omitted was set, but no servers were listed in the room"
-                )
+            servers_in_room = response.servers_in_room
+            if response.members_omitted:
+                if not servers_in_room:
+                    raise InvalidResponseError(
+                        "members_omitted was set, but no servers were listed in the room"
+                    )
 
-            if response.members_omitted and not partial_state:
-                raise InvalidResponseError(
-                    "members_omitted was set, but we asked for full state"
-                )
+                if destination not in servers_in_room:
+                    # `servers_in_room` is supposed to be a complete list.
+                    # Fix things up if the remote homeserver is badly behaved.
+                    servers_in_room = [destination] + servers_in_room
```

Comment:
Should we be making this a set? What if the remote server returns the same thing for `servers_in_room` 100 times?

Is it possible for `destination` to *no longer* be in the room at this point? (I'm guessing no because we just got a response from it?)

Reply:
That's a good question. We'll fail to insert the list into the database, because there's a unique constraint. And the join will fail.

There's nothing stopping `destination` from leaving the room immediately after our join really.
If `destination` leaves the room, we'll try syncing state from other servers in the list.

Reply:
Representing the server list as a set sounds like a good idea, so I did it in 163c68c0be.